dispatch Raft messages in goroutine

tbg opened the pull request:
Prior to this change, Transport.Send() would wait for the RPC client to be
ready, which can block for arbitrary amounts of time, on top of there not
being a timeout after which the call was discarded (potentially forever).

fixes #1073

tbg committed:
dispatch Raft messages in goroutine

Prior to this change, Transport.Send() would wait for the RPC client to be
ready, which can block for arbitrary amounts of time, on top of there not
being a timeout after which the call was discarded (potentially forever).

fixes #1073

petermattis commented on server/raft_transport.go:
```diff
@@ -116,28 +117,42 @@ func (t *rpcTransport) Send(req *multiraft.RaftMessageRequest) error {
-		}()
-		return nil
-	}
+	// NewClient can block, so fire-and-forget from here on.
```
How can `NewClient` block? The first heartbeat is done synchronously in `Client.connect`, but that is run on a separate goroutine.

tschottdorf replied:
Sorry, it can't block. It was the client never getting ready (or returning from a client.Go call, depending on when the remote died).

petermattis commented:
LGTM

tschottdorf commented:
I'll rearrange this to use one goroutine per Node (/Address).

tamird commented on server/raft_transport.go:
```diff
@@ -44,8 +46,7 @@ type rpcTransport struct {
 // newRPCTransport creates a new rpcTransport with specified gossip and rpc server.
-func newRPCTransport(gossip *gossip.Gossip, rpcServer *rpc.Server, rpcContext *rpc.Context) (
-	multiraft.Transport, error) {
+func newRPCTransport(gossip *gossip.Gossip, rpcServer *rpc.Server, rpcContext *rpc.Context) (multiraft.Transport, error) {
```
can you back this change out?

tamird commented:
Can you devise a test for this?

bdarnell commented on server/raft_transport.go:
```diff
@@ -116,28 +117,42 @@ func (t *rpcTransport) Send(req *multiraft.RaftMessageRequest) error {
+		client := rpc.NewClient(addr, nil, t.rpcContext)
+		select {
+		case <-client.Closed:
+			log.Errorf("raft client failed to connect")
```
Add a TODO to introduce some sort of error channel so we can pass the error back to multiraft so it can call ReportUnreachable.

bdarnell commented:
LGTM

tbg committed:
use per-node queue for outgoing raft messages

this should be the best solution: messages are queued per node,
dropping if the queue for the recipient node is full.
a goroutine works on each queue.

tschottdorf commented:
PTAL @bdarnell

tbg committed:
outgoing raft message queue workers shut down after idle period

petermattis commented on server/raft_transport.go:
```diff
@@ -101,43 +115,101 @@ func (t *rpcTransport) Stop(id proto.RaftNodeID) {
+			return
+		case req = <-ch:
+		}
+		if req == nil {
```
How will you ever get a `nil` request here?

tschottdorf replied:
if `ch` is closed. Not that I would expect that to happen, but I like not having to check that.

bdarnell replied:
You can use the `, ok` form when reading from the channel to be explicit that you're checking for a closed channel.

petermattis commented on server/raft_transport.go:
```diff
@@ -101,43 +115,101 @@ func (t *rpcTransport) Stop(id proto.RaftNodeID) {
+		case <-client.Ready:
+		}
+
+		call := client.Go(raftMessageName, protoReq, &proto.RaftMessageResponse{}, nil)
```
You could pull `proto.RaftMessageResponse{}` out of the loop as well. Safe because there is no response.

tschottdorf replied:
sounds good.

petermattis commented on server/raft_transport.go:
```diff
@@ -101,43 +115,101 @@ func (t *rpcTransport) Stop(id proto.RaftNodeID) {
+		}
+
+		deadline := time.After(raftMessageTimeout)
+		client := rpc.NewClient(addr, nil, t.rpcContext)
```
Can the `nodeID -> addr` translation and `client` creation be done outside the loop.

tschottdorf replied:
If you restarted a remote node with a new address, that would mean that the server running this loop would never update, though Gossip does.

tschottdorf replied:
pulled it out as per our discussion. processQueue exits when the client fails anyways.

petermattis commented on server/raft_transport.go:
```diff
@@ -101,43 +115,101 @@ func (t *rpcTransport) Stop(id proto.RaftNodeID) {
+			log.Errorf("could not marshal message: %s", err)
+		}
+
+		nodeID, _ := proto.DecodeRaftNodeID(proto.RaftNodeID(req.Message.To))
```
Won't `proto.RaftNodeID(req.Message.To)` equal the `nodeID` parameter passed to `processQueue`?

tschottdorf replied:
that's correct. I'll pull it out of the loop.

petermattis commented on server/raft_transport.go:
```diff
@@ -101,43 +115,101 @@ func (t *rpcTransport) Stop(id proto.RaftNodeID) {
+		case <-deadline:
+			log.Errorf("call timed out after %s", raftMessageTimeout)
+			continue
+		case <-call.Done:
```
You're serializing raft message sending here. Is it desirable to send multiple raft messages to a node concurrently? If yes, I think we can restructure this just a little bit to pass in our own buffered `done` channel and then use a single select statement inside the loop. But perhaps that is better left for a TODO and a subsequent change.

bdarnell replied:
Yes, it is important for performance to have multiple raft messages to the same node concurrently. I'm OK with making it a subsequent change, though.

tschottdorf replied:
I've given it a shot, PTAL.

petermattis commented:
LGTM

bdarnell commented on server/raft_transport.go:
```diff
@@ -32,8 +33,19 @@ import (
+// raftMessageTimeout specifies the acceptable wait time for a single
+// message to a node to succeed.
+// It is a variable to adapt it for race testing on slow machines.
+var raftMessageTimeout = 1 * time.Second
```
1 second seems aggressive for production too in geographically-distributed deployments. Raft messages may be large (snapshots) and since snapshots are often the first things sent the TCP connection may still be in its slow-start mode.

tschottdorf replied:
The parameter is removed in the most recent commit.

bdarnell commented on server/raft_transport.go:
```diff
@@ -32,8 +33,19 @@ import (
+	raftSendBufferSize = 500
+	// When no message has been sent to a Node for that duration, the
+	// corresponding instance of processQueue will shut down.
+	raftIdleTimeout = time.Minute
```
I'm not sure an idle timeout makes sense. It's cheap to start and stop goroutines so we could stop the processQueue thread as soon as the queue becomes empty. Conversely, it should be affordable to simply keep one processQueue goroutine per peer node for the life of the transport. Running goroutines are cheap and we expect most node-to-node connections to be active most of the time.

I'm concerned about race conditions when a new outgoing message appears just as the old goroutine is shutting down; with an idle timeout this race will occur rarely and be difficult to test. If we shut down the goroutine eagerly we are much more likely to uncover all the bugs; if we never shut it down we avoid the race completely.

tschottdorf replied:
I just didn't want goroutines for nodes which are removed to stay around. And firing up a goroutine with each request (we're doing fire-and-forget, so we might be faster than MultiRaft is feeding our channel) seemed wasteful as well. Operating on the premise that we will always drop the queue on the floor (which might be a good idea anyways, a lot of redundant heartbeats and appends piling up) I don't think this is an issue.
For testing your suggestion to shut down the goroutine eagerly (but then we should probably do it every time) makes sense but in production that seems like something to avoid.

bdarnell commented on server/raft_transport.go:
```diff
@@ -101,43 +115,101 @@ func (t *rpcTransport) Stop(id proto.RaftNodeID) {
+		select {
+		case <-t.rpcContext.Stopper.ShouldStop():
+			return
+		case <-time.After(raftIdleTimeout):
```
This has a race: if we read from the time.After channel, then `ch` is empty at this moment, but `Send` may write something to `ch` before the deferred function fires and discards the channel. Races that simply drop messages aren't crucial since raft will re-send things as necessary, but it would be better to avoid them.

tschottdorf replied:
Whenever the process loop stops, it will discard everything in the queue so we'll drop a lot of messages anyways. I think that's okay given it's MultiRaft and I don't want to introduce additional complexity for the most likely use case (in which the client is fine). We've had more trouble with messages piling up so far.

bdarnell commented:
LGTM

tschottdorf commented on server/raft_transport.go:
```diff
@@ -101,45 +112,106 @@ func (t *rpcTransport) Stop(id proto.RaftNodeID) {
+	ch, ok := t.queues[raftNodeID]
+	if !ok {
+		ch = make(chan *multiraft.RaftMessageRequest, raftSendBufferSize)
+		t.queues[raftNodeID] = ch
```
we could avoid dropping messages at all by Unlock'ing at end of the function and not recreating the channel (we could put a tuple in the map indicating if it's being processed). I'm not sure that dropping messages while a node is down is bad though. Your call.

bdarnell replied:
I don't think that completely solves the race. I agree that dropping messages while a node is down is not a problem. Dropping messages when a node is on the edge of the idle timeout is not ideal but probably not worth a complicated solution. The one thing that would concern me is if there were something that would try to send messages at regular intervals that happened to coincide with the idle timeout (and might therefore hit the race condition repeatedly), but the idle timeout is long enough that it's probably not an issue.

tschottdorf commented:
PTALL (please take a last look)

petermattis commented on server/raft_transport.go:
```diff
@@ -101,45 +112,106 @@ func (t *rpcTransport) Stop(id proto.RaftNodeID) {
+		}
+
+		if !client.IsHealthy() {
+			log.Warningf("client is unhealthy")
```
Add the client address to this log message. Or the node id. Or both.

tschottdorf replied:
done. For all messages.

petermattis commented on server/raft_transport.go:
```diff
@@ -101,45 +112,106 @@ func (t *rpcTransport) Stop(id proto.RaftNodeID) {
+		protoReq.GroupID = req.GroupID
+		var err error
+		if protoReq.Msg, err = req.Message.Marshal(); err != nil {
+			log.Errorf("could not marshal message: %s", err)
```
You probably want to `continue` here. Sending the message can't be right.

tschottdorf replied:
Yup. Changed.

bdarnell commented:
LGTM

tbg committed:
address feedback

- use a buffered channel to collect the responses and log errors,
  so messages are fired rapidly without waiting for them to finish
- removed raftMessageTimeout - it was not necessary any more
- rely on client.IsHealthy()
- factor out some allocations from the loop

tschottdorf commented:
@tamird I've updated #1029 to include tests for this. Feel free to gnaw some of it down, I could use a little help plowing through all those bugs @MBerhault is digging up.
